pkg/ibmvpc/parser.go

```diff
@@ -380,27 +381,58 @@ func (rc *IBMresourcesContainer) getInstancesConfig(
 		vpcConfig := res.Config(vpcUID)
 		vpcConfig.NodeSets = append(vpcConfig.NodeSets, vsiNode)
 		vpcConfig.UIDToResource[vsiNode.ResourceUID] = vsiNode
+		// a VSI can not have both VNIs and NIs.
+		// first we check for VNIs.
+		// the VNIs are listed at the instance.NetworkAttachments.
+		// using that list, we extract the VNIs from the rc.VirtualNIList, which holds all the VNIs of all VSIs
+		for j := range instance.NetworkAttachments {
+			vniObj := vnisObj[*instance.NetworkAttachments[j].ID]
+			err := createNetworkInterface(*vniObj.Name, *vniObj.ID,
+				*instance.Zone.Name, *vniObj.PrimaryIP.Address, *instance.Name, true, vsiNode, len(instance.NetworkAttachments),
+				*vniObj.Subnet.CRN, subnetIDToNetIntf, vpc, vpcConfig)
+			if err != nil {
+				return err
+			}
+		}
+		if len(instance.NetworkAttachments) > 0 {
+			// this VSI has VNIs, we do not check for NIs
+			// notice that in this case, instance.NetworkInterfaces will not be empty,
+			// since each VNI has "NI representation" for backward computability
+			continue
+		}
 		for j := range instance.NetworkInterfaces {
 			netintf := instance.NetworkInterfaces[j]
 			// netintf has no CRN, thus using its ID for ResourceUID
-			intfNode, err := commonvpc.NewNetworkInterface(*netintf.Name, *netintf.ID,
-				*instance.Zone.Name, *netintf.PrimaryIP.Address, *instance.Name, len(instance.NetworkInterfaces), vpc)
+			err := createNetworkInterface(*netintf.Name, *netintf.ID,
+				*instance.Zone.Name, *netintf.PrimaryIP.Address, *instance.Name, false, vsiNode, len(instance.NetworkInterfaces),
+				*netintf.Subnet.CRN, subnetIDToNetIntf, vpc, vpcConfig)
```

Comment:
better pass just `instance` instead of `*instance.Zone.Name`, `*instance.Name` and `len(instance.NetworkInterfaces)`

Reply:
I think it's ok to always use `len(instance.NetworkInterfaces)` because of the guaranteed backwards compatibility.

Reply:
fixed. (there are two calls, one with len(instance.NetworkInterfaces) and the other with len(instance.NetworkAttachments))

Reply:
fixed